### ISSUE TITLE
[PyTorch] High CPU overhead on Grace systems

### ISSUE THREAD

**Issue by Marks101** [NVIDIA/TransformerEngine#2053]
Hello everyone,

we have been investigating low compute throughput for LLM trainings on GH200 and GB200 systems in comparison to H100 DGX (Intel Xeon Platinum 8480C). This already affects 1 GPU trainings and we tracked this down to CPU overhead of Transformer Engine (TE) modules on Grace systems.

To showcase this, I appended a small benchmarking script, that evaluates GPU kernel runtime and CPU overhead of the Linear module from PyTorch and Transformer Engine on GB200, GH200 and H100 DGX.
Our benchmarking does multiple runs and we evaluate the average time per Linear forward pass both on CPU and GPU. In the following run we look at BF16 precision with a hidden size of 2048 and 8192 tokens, these values are typical for 7B LLM trainings:

|       | backend | CPU runtime [µs] | GPU runtime [µs] | estimated throughput [TFLOP/s] |
|-------|---------|------------------|------------------|--------------------------------|
| GB200 | torch   | 37.08            | 41.79            | 1644.22                        |
| GB200 | TE      | 99.66            | 99.73            | 689.08                         |
| GH200 | torch   | 33.17            | 91.64            | 749.86                         |
| GH200 | TE      | 90.53            | 92.14            | 745.80                         |
| H100 DGX | torch   | 13.48            | 100.06           | 686.79                         |
| H100 DGX | TE      | 49.19            | 101.88           | 674.51                         |

So what we see is that:
- The torch Linear has significantly lower CPU overhead in comparison to TE. This is probably expected as TE offers a lot more features.
- On Grace systems, the CPU overhead is almost a factor of two higher in comparison to H100 DGX

This two factors together can severely limit the throughput as execution becomes CPU bound, see for example the GB200 numbers with TE. Although the GPU has 2.5x the FLOP/s, the scenario does not make use of it. Even the GH200 run with TE is very close to be severly CPU bound, especially if you consider that actual training code adds additional CPU instructions.

Generally, increasing the number of tokens minimizes this issue as it increases GPU runtime but CPU overhead is independent of the problem size. But for many scenarios other constraints like memory, global batch size and pipelining considerations limit the number of tokens per forward/backward pass and as such are infeasible. Putting this all together the CPU overhead can severly limit the compute throughput during training.

We already did profiling with nsys and cProfile but were not able to identify any "easy wins". It would be great if you could try to reproduce this and give a first assesment. Hope that you have a few ideas how to improve the situation 😄 

Please let me know if you have further questions or need more details.

**Steps to reproduce**

<details>
  <summary>Benchmark script</summary>

```python
from typing import List
import torch
import time
import argparse

import transformer_engine.pytorch as te


def speedometer(
    module: torch.nn.Module,
    args: List[torch.Tensor],
    timing_iters: int = 500,
    warmup_iters: int = 50,
    num_rounds: int = 5,
) -> float:
    """Measure average run time for a PyTorch module"""
    for _ in range(warmup_iters):
        module(*args)

    gpu_times = []
    cpu_times = []
    for round_idx in range(num_rounds):
        start = torch.cuda.Event(enable_timing=True)
        end = torch.cuda.Event(enable_timing=True)
        torch.cuda.synchronize()
        start.record()
        cpu_start = time.time()
        for _ in range(timing_iters):
            module(*args)
        cpu_end = time.time()
        end.record()
        torch.cuda.synchronize()
        gpu_elapsed = start.elapsed_time(end)
        cpu_elapsed = (cpu_end - cpu_start) * 1000
        gpu_times.append(gpu_elapsed)
        cpu_times.append(cpu_elapsed)
        print(
            f"Round {round_idx+1}/{num_rounds}: GPU {gpu_elapsed/timing_iters*1000:.2f} µs, CPU {cpu_elapsed/timing_iters*1000:.2f} µs"
        )
    print(f"Average GPU time over {num_rounds} rounds: {sum(gpu_times)/(num_rounds*timing_iters)*1000:.2f} µs")
    print(f"Average CPU time over {num_rounds} rounds: {sum(cpu_times)/(num_rounds*timing_iters)*1000:.2f} µs")

    return sum(gpu_times) / num_rounds


def main():
    parser = argparse.ArgumentParser(description="Benchmark torch.nn.Linear performance and CPU overhead.")
    parser.add_argument("--hidden_size", type=int, default=2048, help="Hidden size")
    parser.add_argument("--seq_length", type=int, default=8192, help="Sequence length")
    parser.add_argument("--warmup", type=int, default=500, help="Number of warmup iterations")
    parser.add_argument("--timing_iters", type=int, default=500, help="Number of timing iterations per round")
    parser.add_argument("--num_rounds", type=int, default=3, help="Number of timing rounds")
    parser.add_argument(
        "--backend", type=str, choices=["torch", "te"], default="te", help="Linear backend: torch or te"
    )
    args = parser.parse_args()

    x = torch.randn((args.seq_length, args.hidden_size), dtype=torch.bfloat16, device="cuda", requires_grad=True)
    if args.backend == "torch":
        model = torch.nn.Linear(args.hidden_size, args.hidden_size, bias=False).to(torch.bfloat16).cuda()
    else:
        model = te.Linear(args.hidden_size, args.hidden_size, bias=False, device="cuda").to(torch.bfloat16)
    avg_gpu_time_per_round = speedometer(
        model, [x], timing_iters=args.timing_iters, warmup_iters=args.warmup, num_rounds=args.num_rounds
    )

    total_ops = 2 * args.hidden_size * args.hidden_size * args.seq_length * args.timing_iters

    tflops = total_ops / avg_gpu_time_per_round / 1e9
    print(f"Estimated TFLOP/s: {tflops:.2f}")


if __name__ == "__main__":
    main()
```
</details>

The numbers above can be reproduced by executing the given benchmarking script with:

```bash
python benchmark_linear_cpu_overhead.py --backend torch
python benchmark_linear_cpu_overhead.py --backend te
```

**Environment overview**

- Docker container: `nvcr.io/nvidia/pytorch:25.06-py3`
- GB200: NVIDIA GB200 Superpod with NVL72
- GH200: NVIDIA GH200 480GB
- H100 DGX: NVIDIA H100 80GB HBM3

**Comment by ptrendx** [comment]
Sigh. Yes, we are very much aware about the general issue of the CPU overhead being higher than the native pyTorch operators (although were not quite aware about impact being higher on Grace, will look into that).

As you said, there are no easy wins there really (apart from trying to utilize CUDA graphs, although in the case of training and especially in the context of pipeline parallelism I would not really call it easy :-) ) and a lot of the issues comes fundamentally from Python and also the way pyTorch treats the autograd functions. As part of the TE 2.0 refactoring we already moved more of the logic into the C++ layer (since even such simple things as calling a view of the tensor on the Python side results in ~1us of the additional CPU time and as you can see from the measurements, 1 us is actually quite a lot). We decided against moving more of the logic to C++  to still enable other people to hack inside the modules without too many barriers. One thing that we did not try yet but would like to try, is utilizing Cython to limit the Python overheads as they are the dominating factor. Without dealing with them, the CPU overhead improvements which we will be able to do will be mostly incremental unfortunately :-(.

All of that said though, in typical training we did not really see that issue pop up too much as typically the tensors were larger than the values you provided (especially the hidden dimension being 2k seems quite small, I typically saw hidden dim being on the order of 8k-12k). Could you provide more info on the model you are seeing the issue with?

**Comment by Marks101** [comment]
Hi @ptrendx thanks for answering so quickly 😄 

> All of that said though, in typical training we did not really see that issue pop up too much as typically the tensors were larger than the values you provided (especially the hidden dimension being 2k seems quite small, I typically saw hidden dim being on the order of 8k-12k). Could you provide more info on the model you are seeing the issue with?

Generally we did not have issues with H100 DGX before as well. For Grace that's different now. A hidden dimension of 2048 is indeed in the domain of smaller dense models. But for example the Qwen3-30B-A3B has this hidden size as well. LLMs in the range of 7B typically have hidden dimensions of 4096. But when trained with a micro batch size of 1 and a sequence lenght of 4096, we have encountered CPU boundedness as well on Grace.

> As part of the TE 2.0 refactoring we already moved more of the logic into the C++ layer (since even such simple things as calling a view of the tensor on the Python side results in ~1us of the additional CPU time and as you can see from the measurements, 1 us is actually quite a lot).

torch `TensorBase.view()` is indeed a good example. The profiles I took with cProfile indicate, that it takes around 5µs on Grace, on our H100 DGX it is around 1.4µs. 

Would be really great if you could do some benchmarks on Grace as well. Looking forward to your results 🤗

**Comment by Marks101** [comment]
One thing I like to add: I mentioned in my last post that we have seen CPU boundedness on a 7B LLM on Grace. This was on GH200 and in a setup using FP8. As FP8 reduces GPU kernel runtime, this makes it easier to go into CPU boundedness ...

**Comment by zhongbozhu** [comment]
Thanks for your observation! CPU overhead is indeed a problem we want to solve; Immediate fix is to enable cuda graph in the training process. Using larger problem shapes will also help in this process. 

It's true that TE with all the low precision support and cross generation arch support is adding a lot of extra logic that brings CPU overhead. For example, in the case of FP8, checking for TN gemm layout is not needed for Hopper only arch, but now we need to launch CUDA apis to check the supported gemm layout (we can cache this check for sure but then we also need to decide when to invalidate the cache): https://github.com/NVIDIA/TransformerEngine/blob/3fc1e4bf8d46850f215698b39b5625310194bab9/transformer_engine/common/transformer_engine.cpp#L645 

Note that the "GPU runtime" measurement in your chart is not exactly the GPU kernel runtime because the time measured is also affected by CPU. From nsys profile, the kernels in `te.Linear` and `torch.nn.linear` are equally fast, because they all use the same cublas BF16 gemm. 

For fwd only comparison, you might also need `torch.no_grad` for TE to skip some logic since it will try to prepare for backward. TE checks for whether grad is needed and decide whether it will cache tensors for backward, this step is mainly CPU.

<details>

<summary>example: </summary>

```python
def speedometer(
    module: torch.nn.Module,
    args: List[torch.Tensor],
    timing_iters: int = 500,
    warmup_iters: int = 50,
    num_rounds: int = 5,
) -> float:
    """Measure average run time for a PyTorch module"""
    with torch.no_grad():
        for _ in range(warmup_iters):
            module(*args)

        gpu_times = []
        cpu_times = []
        for round_idx in range(num_rounds):
            start = torch.cuda.Event(enable_timing=True)
            end = torch.cuda.Event(enable_timing=True)
            torch.cuda.synchronize()
            start.record()
            cpu_start = time.time()
            for _ in range(timing_iters):
                module(*args)
            cpu_end = time.time()
            end.record()
            torch.cuda.synchronize()
            gpu_elapsed = start.elapsed_time(end)
            cpu_elapsed = (cpu_end - cpu_start) * 1000
            gpu_times.append(gpu_elapsed)
            cpu_times.append(cpu_elapsed)
            print(
                f"Round {round_idx+1}/{num_rounds}: GPU {gpu_elapsed/timing_iters*1000:.2f} µs, CPU {cpu_elapsed/timing_iters*1000:.2f} µs"
            )
        print(f"Average GPU time over {num_rounds} rounds: {sum(gpu_times)/(num_rounds*timing_iters)*1000:.2f} µs")
        print(f"Average CPU time over {num_rounds} rounds: {sum(cpu_times)/(num_rounds*timing_iters)*1000:.2f} µs")

        return sum(gpu_times) / num_rounds
```
</details>

**Comment by zhongbozhu** [comment]
One more thing to add is, I found hidden size 3072 from your script is roughly the threshold value for te.Linear is matching torch linear, which means that we are not bounded by CPU in this case. However, there are still many places to optimize to bring down the threshold as much as possible, things like removing extra logic for a cleaner path for bf16 baseline, improving pybind perf, using the the WIP te.Sequential API to put as much as kernels into minimal amount of pytorch autograd classes. Current TE linear is calling customed stateless fwd bwd torch.autograd.Function inside a stateful torch nn Module, but even that is not fast enough for GB200 because two autograd classes are involved in CPU while GPU finishes too fast. 

Moving to FP8, hidden size 4k is roughly the threshold for the fp8cast+gemm to outperform bf16, so CPU overhead issue is not as severe.

Another thing to consider is that, in training, there are backward passes, backward linear has two GEMMs to compute, one for wgrad and one for dgrad, so while the CPU is launching kernels for the forward pass, GPU is still busy with the previous backward pass, so the CPU overhead will not be exposed in many training workloads.

**Comment by Marks101** [comment]
Hi @zhongbozhu ,

thank you very much vor your detailed comments and insights.

> Note that the "GPU runtime" measurement in your chart is not exactly the GPU kernel runtime because the time measured is also affected by CPU. From nsys profile, the kernels in te.Linear and torch.nn.linear are equally fast, because they all use the same cublas BF16 gemm.

Yes, that's right. I am aware of this. The CPU kernel measurement breaks down when the benchmark is CPU bound. But it was never the intent of this script to give an exact measurement of kernel runtime. It was just to show when things get CPU bound.

> For fwd only comparison, you might also need torch.no_grad for TE to skip some logic since it will try to prepare for backward. TE checks for whether grad is needed and decide whether it will cache tensors for backward, this step is mainly CPU.

You are totally right. My benchmarking script is showcasing an extreme scenario. It would be fairer if the backward pass is included in the benchmark. Because as you suggested in your next message:

> Another thing to consider is that, in training, there are backward passes, backward linear has two GEMMs to compute, one for wgrad and one for dgrad, so while the CPU is launching kernels for the forward pass, GPU is still busy with the previous backward pass, so the CPU overhead will not be exposed in many training workloads.

we have seen this as well in nsys profiles: the backward pass has less CPU overhead and so it can make up some time for the upfollowing forward pass.


What would you recommend as best way to determine if a training is CPU bound? I guess nsys profiling? Which nsys settings are most appropriate in order to avoid that the profiling introduces additional CPU overhead which makes it hard to detect CPU boundedndess?

I will check if I can provide more insights based on a nsys profile.

> However, there are still many places to optimize to bring down the threshold as much as possible,

Do you have a roadmap for these improvements? Any specific TE version you are targeting?

**Comment by zhongbozhu** [comment]
You can find some of the nsys command examples here: https://github.com/NVIDIA/TransformerEngine/blob/d770886f02b6c43da1afb65d19487a44cbb8cc88/benchmarks/linear/benchmark_grouped_linear.py#L18

Typically when you are CPU bounded, you will find the row "CUDA HW" mostly blank instead of blue. You can click open this row to see what kernel is being run in the blue parts. 

You can also check the row with called NVTX. This is basically CPU side and TE has already annotated each component in the linear layer's forward and backward pass. You can use NVTE_NVTX_ENABLED env variable to enable more of those annotations. In many cases, being CPU bounded is caused by having many tiny PyTorch kernels or many torch.empty memory allocations, those issues can be precisely captured by nsys. For TE, most of these optimizations are already done, but CPU overhead can also be caused by other components of the training stack.

To mitigate the CPU overhead really depends on your training framework. If your training framework is MegatronLM, then it already has flags to capture CUDA graph. For MegatronLM, it's called `--enable-cuda-graph`. TE on its own can also capture CUDA graph, this is helpful for many MOE training workloads where a full capture is not possible, so you can use the `--external-cuda-graph` flag from MegatronLM to let TE capture the CUDA graph for itself, and you can also specify the scope you want to partially capture: eg. only capture for attention. If you use other training frameworks but integrated TE, then you can explore the TE's offerings of CUDA graph API to capture yourself. 

As for the roadmap for improving CPU overhead in general, I think we would basically try to make cuda graph easier to use, and also try to move some more logic away from pytorch into C++ as well as caching some repetitive CUDA API calls (eg. for checking sm architecture). Mostly micro-optimizations gradually coming into the Lib and accumulate them overtime.

**Comment by zhongbozhu** [comment]
CPU overhead should have been greatly reduced as of today, which basically means that when both running BF16 dense linear without any fancy features, TELinear is nearly the same with vanilla torch linear. Some more PRs are trying to optimize CPU side further. 

On the other hand, enabling CUDA Graph in the training is still crucial especially for MOE. As of today, partial capture for dense components is fully supported in Mcore, while full iteration cuda graph including MOE layers is also being actively worked on. Stay tuned :)
